packages/react-aria/src/hooks/useARIAButton.ts

```diff
@@ -1,99 +1,104 @@
+  const { disabled, disabledFocusable, onClick, onKeyDown, onKeyUp, tabIndex } = (shorthand ||
+    {}) as ARIAButtonShorthandProps;
 
   const onClickHandler: ARIAButtonShorthandProps['onClick'] = useEventCallback(ev => {
-    if (disabled) {
+    if (disabled || disabledFocusable) {
       ev.preventDefault();
       ev.stopPropagation();
     } else {
-      if (typeof onClick === 'function') {
-        onClick(ev);
-      }
+      onClick?.(ev);
     }
   });
 
   const onKeyDownHandler: ARIAButtonShorthandProps['onKeyDown'] = useEventCallback(ev => {
-    if (typeof onKeyDown === 'function') {
-      onKeyDown(ev);
-    }
+    onKeyDown?.(ev);
+
     if (ev.isDefaultPrevented()) {
       return;
     }
-    if (disabled && (ev.key === Enter || ev.key === Space)) {
+
+    const key = ev.key;
+
+    if ((disabled || disabledFocusable) && (key === Enter || key === Space)) {
       ev.preventDefault();
       ev.stopPropagation();
       return;
     }
-    if (ev.key === Space) {
+
+    if (key === Space) {
       ev.preventDefault();
       return;
     }
+
     // If enter is pressed, activate the button
-    else if (ev.key === Enter) {
+    else if (key === Enter) {
       ev.preventDefault();
       ev.currentTarget.click();
     }
   });
 
   const onKeyupHandler: ARIAButtonShorthandProps['onKeyUp'] = useEventCallback(ev => {
-    if (typeof onKeyUp === 'function') {
-      onKeyUp(ev);
-    }
+    onKeyUp?.(ev);
+
     if (ev.isDefaultPrevented()) {
       return;
     }
-    if (disabled && (ev.key === Enter || ev.key === Space)) {
+
+    const key = ev.key;
+
+    if ((disabled || disabledFocusable) && (key === Enter || key === Space)) {
       ev.preventDefault();
       ev.stopPropagation();
       return;
     }
-    if (ev.key === Space) {
+
+    if (key === Space) {
       ev.preventDefault();
       ev.currentTarget.click();
     }
   });
 
   if (shorthand) {
+    // If a <button> tag is to be rendered we just need to set disabled and aria-disabled correctly
     if (shorthand.as === 'button' || shorthand.as === undefined) {
-      return shorthand; // there's nothing to be done if as prop === 'button' or undefined
+      shorthand.disabled = disabled && !disabledFocusable;
+      shorthand['aria-disabled'] = disabledFocusable;
     }
 
-    if (!shorthand.hasOwnProperty('role')) {
-      shorthand.role = 'button';
+    // If an <a> tag is to be rendered we have to remove disabled and set aria-disabled, role and tabIndex as well as
+    // some event handlers.
+    else {
+      delete shorthand.disabled;
+      shorthand['aria-disabled'] = disabled || disabledFocusable;
+      shorthand.onClick = onClickHandler;
+      shorthand.onKeyDown = onKeyDownHandler;
+      shorthand.onKeyUp = onKeyupHandler;
+      shorthand.role = shorthand.role ?? 'button';
+      shorthand.tabIndex = disabled && !disabledFocusable ? undefined : tabIndex ?? 0;
```

Comment:
I half wonder if there's a way to mandate an `href` attribute if someone wants `as: a`, since otherwise it's effectively `as: span`, and then we don't need to do the weird keydown vs. keyup enter key dance. And also then we wouldn't need to specify tabindex unless it's disabled :)

Reply:
I think this is something that we should explore, but this PR is very large as is, so I'll leave it till later to investigate this.